plutus-core/plutus-core/test/Pretty/Golden/Readable/StdLib/Data/List/FoldrList.golden

```diff
@@ -25,15 +25,15 @@
       {r}
       (\(rec : list a -> r) (xs : list a) ->
          (/\a ->
-            \(x : list a) ->
+            \(xs : list a) ->
               /\r ->
                 \(z : r) (f : a -> list a -> r) ->
-                  chooseList
+                  caseList
                     {a}
                     {unit -> r}
-                    x
+                    xs
                     (\(u : unit) -> z)
-                    (\(u : unit) -> f (headList {a} x) (tailList {a} x))
+                    (\(x : a) (xs' : list a) (u : unit) -> f x xs')
```

Comment:
If i understand correctly, we would need to allow higher-order builtins?

Reply:
That is, is `caseList : all a b. list a -> (unit -> b) -> (a -> list a -> unit -> b)` ?
Why is the unit in the last argument needed?

Reply:
> If i understand correctly, we would need to allow higher-order builtins?

Yes, but only in the metatheory, the implementation supports arbitrary Plutus types in type signatures of builtins (i.e. the metatheory does it better).

> is `caseList : all a b. list a -> (unit -> b) -> (a -> list a -> unit -> b)` ?

No, the type of `caseList` doesn't have any units, even though they're used inside. They probably shouldn't be used inside, like with the `caseData` definition discussion below. Note that this is the nonsensical stuff from the `stdlib`, so don't think too hard about it, the actual builtin (confusingly named the same) is defined elsewhere, here it's only tested.